File: week-7/dahyen0o/택배-배달과-수거하기.java
import java.util.*;

class Solution {
    public long solution(final int cap, final int n, int[] deliveries, int[] pickups) {
        long answer = 0;
        
        int d = deliveries.length - 1;
        int p = pickups.length - 1;
        
        while(d >= 0 && p >= 0) {
            // cap으로 가능한 배달과 수거 진행
            while(d >= 0 && deliveries[d] == 0) {
                d--;
            }
            while(p >= 0 && pickups[p] == 0) {
                p--;
            }
            final int endPoint = Math.max(d, p) + 1;
            answer += endPoint * 2;
            
            int dCap = cap;
            while(d >= 0) {
                if(dCap > deliveries[d]) {
                    dCap -= deliveries[d--];
                } else {
                    deliveries[d] -= dCap;
                    break;
                }
            }
            
            int pCap = cap;
            while(p >= 0) {
                if(pCap > pickups[p]) {
                    pCap -= pickups[p--];
                } else {
                    pickups[p] -= pCap;
                    break;
                }
            }
        }
                
        if(d >= 0) { // pickup ended
            while(d >= 0) {
                while(d >= 0 && deliveries[d] == 0) {
                    d--;
                }
                answer += (d + 1) * 2;

                int dCap = cap;
                while(d >= 0) {
                    if(dCap > deliveries[d]) {
                        dCap -= deliveries[d--];
                    } else {
                        deliveries[d] -= dCap;
                        break;
                    }
                }
            }
        } else { // deliveries ended
            while(p >= 0) {
                while(p >= 0 && pickups[p] == 0) {
                    p--;
                }
                answer += (p + 1) * 2;

                int pCap = cap;
                while(p >= 0) {
                    if(pCap > pickups[p]) {
                        pCap -= pickups[p--];
                    } else {
                        pickups[p] -= pCap;
                        break;
                    }
                }
            }
        }
        
        return answer;
    }
    
}

File: week-7/rimrim990/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
  # 배달해야 할 집
  deliver = -1
  for i in range(n-1, -1, -1):
    if deliveries[i] > 0:
      deliver = i
      break

  # 수거해야 할 집
  pick = -1
  for i in range(n-1, -1, -1):
    if pickups[i] > 0:
      pick = i
      break

  # 투 포인터
  answer = 0
  while deliver >= 0 or pick >= 0:
    answer += 2 * (max(pick, deliver) + 1)

    # 배달하기
    hand = 0
    while deliver >= 0 and hand <= cap:
      if hand + deliveries[deliver] <= cap:
        hand += deliveries[deliver]
        deliver -= 1
        continue

      available = cap - hand
      deliveries[deliver] -= available
      break

    # 수거하기
    hand = 0
    while pick >= 0 and hand <= cap:
      if hand + pickups[pick] <= cap:
        hand += pickups[pick]
        pick -= 1
        continue

      available = cap - hand
      pickups[pick] -= available
      break

  return answer